go through has_item

mattman107 opened the pull request.

mattman107 committed:
go throuh has_item

mattman107 commented on worlds/oot_soh/LogicHelpers.py:
```diff
@@ -118,56 +118,81 @@ def has_item(item: Items | Events | Enum, bundle: tuple[CollectionState, Regions
+        case Items.BOTTLE_WITH_RED_POTION:
+            return has_bottle(bundle) # TODO Should we check if we can buy red potion?
+        case Items.EMPTY_BOTTLE:
+            return has_bottle(bundle)
```
Technically all of the bottle checks in Ship only do `has_bottle` check here, but since we have some events to determine if we have access to certain things I added them.

mattman107 commented:
#104

mattman107 committed:
Add Epona

mattman107 commented:
Probably would be easier to merge #145 first and then I can combine these changes into that.

mattman107 committed:
Update tuple to set

ScipioWright commented on worlds/oot_soh/LogicHelpers.py:
```diff
@@ -113,61 +113,88 @@ def has_item(item: Items | Events | Enum, bundle: tuple[CollectionState, Regions
-                               Items.PROGRESSIVE_BOMBCHU.value), world.player)
+    if item in {Items.PROGRESSIVE_BOMBCHU, Items.BOMBCHUS_5, Items.BOMBCHUS_10, Items.BOMBCHUS_20}:
+        return (state.has_any({Items.BOMBCHUS_5.value, Items.BOMBCHUS_10.value, Items.BOMBCHUS_20.value,
+                               Items.PROGRESSIVE_BOMBCHU.value}, world.player)
```
It's ever so slightly faster to use tuples here rather than sets. But if this is for readability, then that's fine

mattman107 replied:
The match case was because I preferred the look of it. 

The sets were because a quick search said it was better for membership testing.

After some more research I now understand why both of these perform at least a bit worse in this scenario. So I'm planning on rewriting my match back to an else if tree and replace the sets with tuples again. 

Thank you for bringing this to my attention and prompting me to do some more research. 👍

mattman107 replied:
When rewriting the if else I plan on optimizing it so what we check the most is at the top.

ScipioWright reviewed (commented):
What's the purpose of changing this from an if else tree to a match case?

mattman107 committed:
Remove sets. We don't get any benefit from them at this scale.

Update match back to chain of if statements. Ordered if statements by most likely to least.

mattman107 committed:
Add fishing pole

mattman107 committed:
add case for adult trade options

aMannus commented on worlds/oot_soh/LogicHelpers.py:
```diff
@@ -112,62 +112,131 @@ def has_item(item: Items | Events | Enum, bundle: tuple[CollectionState, Regions
-    if item in (Items.PROGRESSIVE_BOMBCHU, Items.BOMBCHUS_5, Items.BOMBCHUS_10, Items.BOMBCHUS_20):
+
+    if item == Items.HOOKSHOT: # 112
+        return state.has(Items.PROGRESSIVE_HOOKSHOT.value, player)
```
What's the commented out number for?

mattman107 replied:
Those numbers are the total amount of times this item is checked for in the code base currently. It is a combination of the calls from can_use and has_item. I used it to help order these from most likely to least likely for efficiency. I can remove them.

mattman107 replied:
Removed and reordered as it was determined sticking to ships order was more beneficial.

aMannus commented on worlds/oot_soh/LogicHelpers.py:
```diff
@@ -112,62 +112,131 @@ def has_item(item: Items | Events | Enum, bundle: tuple[CollectionState, Regions
+        return state.has(Items.STRENGTH_UPGRADE.value, player, 3)
+    
+    if item == Items.EPONA: # 5
+        return state.has(Events.FREED_EPONA, player)
```
This needs to check for can_use epona's song and adult. FREED_EPONA is only relevant for doing the obstacle course I'm pretty sure.

mattman107 replied:
According to our Talon Race event:
```    
 (EventLocations.LLR_TALON_RACE, Events.FREED_EPONA, lambda bundle: (has_item(Items.CHILD_WALLET, bundle) or world.options.skip_epona_race.value == 1) and can_play_song(Items.EPONAS_SONG, bundle) and is_adult(bundle) and at_day(bundle)),
```

So we have already checked for this. Just seeing if we completed the event is enough.

aMannus commented on worlds/oot_soh/LogicHelpers.py:
```diff
@@ -112,62 +112,131 @@ def has_item(item: Items | Events | Enum, bundle: tuple[CollectionState, Regions
+        return state.has(Items.PROGRESSIVE_OCARINA.value, player, 2)
+    
+    # Should Items with 0 calls even exist?
+
```
I'd keep them for now, maybe they get introduced later down the line in Ship.

aMannus commented:
Also pretty sure https://github.com/aMannus/Archipelago/pull/145 is going to change this quite a lot as it removes a lot of the direct comparisons.

mattman107 commented:
> Also pretty sure #145 is going to change this quite a lot as it removes a lot of the direct comparisons.

Definitely does. I think that one should be merged first and then I can make the required changes. I have a version of this that is fixed to work with #145 already just not currently applied to this PR.

aMannus commented:
I just merged #145 so feel free to swap over

mattman107 committed:
Merge branch 'soh' of https://github.com/aMannus/Archipelago into has_item

mattman107 committed:
Requested changes

aMannus reviewed (approved):
I'm going to merge this as is but I am making an issue for the shop items, as I'm pretty sure we're not going to be using events to handle these but have to directly compare to the shop item itself.
